Advanced Options for Single Crystal Elastic mode of DNS Reduction GUI

koshchii opened the pull request:
### Description of work

This PR contains advanced data reduction options selection capabilities for the Single Cristal Elastic mode of DNS Reduction GUI. Complete integration of the Single Cristal Elastic mode into the DNS Reduction GUI will be implemented in several steps (in order to ensure reasonable PR sizes and reduce the workload for the reviewers), and this PR represents the 5th step of the integration process.

#### Summary of work
Fixes Part 5 of #36407, as well as a bug reported in #37712.


### To test:

- Download test data: [Single_Crystal_Elastic.zip](https://github.com/mantidproject/mantid/files/13751286/Single_Crystal_Elastic.zip)
- Open the reduction GUI [Interfaces->Direct->DNS Reduction]
- Make sure that the "Single Crystal Elastic" mode opens by default. Otherwise switch to this mode, by selecting Tools -> Change Mode -> Single Crystal Elastic
- Make sure that the corresponding documentation is up-to-date (run `cmake --build . --target docs-html` in build folder to create docs)
- Browse to the directory where the test data are located using the "Paths" tab of the GUI
- Load and select the sample data for reduction using the "Data" tab
- Click on the "Options" tab and and make sure that various option selection fields work as expected, and the selected options are then applied to the data once the data reduction script is generated using the "Script Generator" tab 
- Reduce the data and generate a data reduction script using the "Script Generator" tab
- Open the "Plotting" tab and make sure that the images are generated. Advanced plotting functionality is not implemented at this stage and will be activated in a follow-up PR. 

*This does not require release notes* because **this PR contains limited functionality.** Release notes for the Single Cristal Elastic mode will be included in a subsequent PR, once all components for implementing the data reduction workflow are uploaded. 

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

koshchii committed:
Add advanced data reduction options selection capabilities

koshchii committed:
Improve binning-related options capabilities

koshchii committed:
Fix bug with expand_all/unexpand_all buttons

koshchii committed:
Update unit tests

Paraquat reviewed (commented):
General observations:

1. If I click on the `Generate script` button under the `Script generator` tab, then paste the script into the main window of Mantid, then run it, the data visualised under the `Plotting` tab is no longer meaningful.
![image](https://github.com/user-attachments/assets/7126203a-f4ab-433c-ac96-715041600d21)

2. This button should be labelled `Generate and run script`. I did not realise that it was actually running the script.
![image](https://github.com/user-attachments/assets/fa14ac17-746c-418e-85c0-5efb11e39902)

3. If I don't click on the `Options` tab during the workflow, I get an arcane message about not setting the binning, when all I needed to do was click on the `Options` tab. This does not seem like user-friendly behaviour.

4. This widget is a bit counter-intuitive. If I edit `2theta min` the hit tab, I'm taken to the `omega min` box instead of `2theta max`. This may just be a personal thing, but I feel like the the 2theta controls should be in a separate widget from the omega controls because it's a little difficult to visually parse. 
![image](https://github.com/user-attachments/assets/c8970737-2bae-4e82-aebe-6ef5ba0143cf)

koshchii committed:
Add changes from code review

koshchii commented:
Comments to the observations with screenshots:
>     1. If I click on the `Generate script` button under the `Script generator` tab, then paste the script into the main window of Mantid, then run it, the data visualised under the `Plotting` tab is no longer meaningful.
1. This is, in principle, not the way how a user is expected to run the workflow, but this behavior should be fixed in future. I added a corresponding to-do-item in issue #36064.
>     2. This button should be labelled `Generate and run script`. I did not realise that it was actually running the script.
2. Thanks, that's fixed now.
>     3. If I don't click on the `Options` tab during the workflow, I get an arcane message about not setting the binning, when all I needed to do was click on the `Options` tab. This does not seem like user-friendly behaviour.
3. This behavior should be investigated and an optimal solution should be found and implemented. I added a corresponding to-do-item in issue #36064.
>     4. This widget is a bit counter-intuitive. If I edit `2theta min` the hit tab, I'm taken to the `omega min` box instead of `2theta max`. This may just be a personal thing, but I feel like the the 2theta controls should be in a separate widget from the omega controls because it's a little difficult to visually parse.
4. Thanks, I think it makes sense. It is fixed now.

sf1919 commented:
As this PR is going to be replaced by another one thats rebased on main I won't re-run any tests as part of the Linux switchover.

koshchii commented:
Closing this because it is replaced by #39310

Paraquat reviewed (commented):
Looks good now

koshchii commented on qt/python/mantidqtinterfaces/mantidqtinterfaces/dns_single_crystal_elastic/helpers/converters.py:
```diff
@@ -9,29 +9,11 @@
-    s13 = a * b**2 * c * (cos_gamma * cos_alpha - cos_beta)
-    inverse_d_square = 1.0 / vol**2 * (s11 * h**2 + s22 * k**2 + s33 * l**2 + 2.0 * s12 * h * k + 2 * s23 * k * l + 2 * s13 * h * l)
-    d = sqrt(1.0 / inverse_d_square)
+    d = round(UnitCell.d(unit_cell, h, k, l), 5)
```
Right, it shouldn't be here. I moved the rounding to the script_generator file, which generates a script that is exposed to a user and 5 decimal places is sufficient there
